lib/ii.c

```diff
@@ -15211,7 +15213,42 @@ grn_ii_builder_flush_term(grn_ctx *ctx, grn_ii_builder *builder,
       }
       return ctx->rc;
     }
-    global_tid = grn_table_add(ctx, builder->ii->lexicon, key, key_size, NULL);
+    /* Don't normalize key. */
+    switch (builder->ii->lexicon->header.type) {
+    case GRN_TABLE_PAT_KEY :
+      GRN_TABLE_LOCK_BEGIN(ctx, builder->ii->lexicon) {
+        global_tid = grn_pat_add(ctx,
+                                 (grn_pat *)(builder->ii->lexicon),
+                                 key,
+                                 key_size,
+                                 NULL,
+                                 NULL);
+      } GRN_TABLE_LOCK_END(ctx, builder->ii->lexicon);
+      break;
+    case GRN_TABLE_DAT_KEY :
+      GRN_TABLE_LOCK_BEGIN(ctx, builder->ii->lexicon) {
+        global_tid = grn_dat_add(ctx,
+                                 (grn_dat *)(builder->ii->lexicon),
+                                 key,
+                                 key_size,
+                                 NULL,
+                                 NULL);
+      } GRN_TABLE_LOCK_END(ctx, table);
```

Comment:
Isn't it `builder->ii->lexicon` instead of `table`?

```suggestion
      } GRN_TABLE_LOCK_END(ctx, builder->ii->lexicon);
```

Reply:
Good catch!

OK. I remove the second argument from `GRN_TABLE_LOCK_END()` because it's an internal macro. We can restore it without braking compatibility when we need it.